File: packages/core/src/parsing/parserHelpers.ts
import { PrintItemIterable, PrintItem, Signal, PrintItemKind, Info } from "../types";

export namespace parserHelpers {
    export function* withIndent(item: PrintItemIterable): PrintItemIterable {
        yield Signal.StartIndent;
        yield* item;
        yield Signal.FinishIndent;
    }

    export function* newlineGroup(item: PrintItemIterable): PrintItemIterable {
        yield Signal.StartNewLineGroup;
        yield* item;
        yield Signal.FinishNewLineGroup;
    }

    export function* prependToIterableIfHasItems<T>(iterable: Iterable<T>, ...items: T[]) {
        let found = false;
        for (const item of iterable) {
            if (!found) {
                yield* items;
                found = true;
            }
            yield item;
        }
    }

    export function* toPrintItemIterable(printItem: PrintItem): PrintItemIterable {
        yield printItem;
    }

    export function* surroundWithNewLines(item: PrintItemIterable): PrintItemIterable {
        yield Signal.NewLine;
        yield* item;
        yield Signal.NewLine;
    }

    /**
     * Reusable function for parsing a js-like single line comment (ex. // comment)
     * @param rawCommentValue - The comment value without the leading two slashes.
     */
    export function parseJsLikeCommentLine(rawCommentValue: string) {
        const nonSlashIndex = getFirstNonSlashIndex();
        const startTextIndex = rawCommentValue[nonSlashIndex] === " " ? nonSlashIndex + 1 : nonSlashIndex;
        const commentValue = rawCommentValue.substring(startTextIndex).trimRight();
        const prefix = "//" + rawCommentValue.substring(0, nonSlashIndex);

        return prefix + (commentValue.length > 0 ? ` ${commentValue}` : "");

        function getFirstNonSlashIndex() {
            for (let i = 0; i < rawCommentValue.length; i++) {
                if (rawCommentValue[i] !== "/")
                    return i;
            }

            return rawCommentValue.length;
        }
    }

    export function createInfo(name: string): Info {
        return {
            kind: PrintItemKind.Info,
            name
        };
    }

    /**
     * Takes a string that could contain tabs or newlines and breaks it up into
     * the correct newlines and tabs.
     * @param text - Text to break up.
     */
    export function* parseRawString(text: string): PrintItemIterable {
        let hasIgnoredIndent = false;

        const lines = text.split(/\r?\n/);
        for (let i = 0; i < lines.length; i++) {
            if (i > 0) {
                if (!hasIgnoredIndent) {
                    yield Signal.StartIgnoringIndent;
                    hasIgnoredIndent = true;
                }

                yield Signal.NewLine;
            }
            yield* parseLine(lines[i]);
        }

        if (hasIgnoredIndent)
            yield Signal.FinishIgnoringIndent;

        function* parseLine(line: string): PrintItemIterable {
            const parts = line.split("\t");
            for (let i = 0; i < parts.length; i++) {
                if (i > 0)
                    yield Signal.Tab;
                yield* parts[i];
            }
        }
    }
}

File: packages/rust-core/src/parser_helpers.rs
use super::print_items::*;

pub fn surround_with_new_lines(item: PrintItems) -> PrintItems {
    let mut items = PrintItems::new();
    items.push_signal(Signal::NewLine);
    items.extend(item);
    items.push_signal(Signal::NewLine);
    items
}

pub fn with_indent(item: PrintItems) -> PrintItems {
    with_indent_times(item, 1)
}

pub fn with_indent_times(item: PrintItems, times: u32) -> PrintItems {
    let mut items = PrintItems::new();
    for _ in 0..times { items.push_signal(Signal::StartIndent); }
    items.extend(item);
    for _ in 0..times { items.push_signal(Signal::FinishIndent); }
    items
}

pub fn with_no_new_lines(item: PrintItems) -> PrintItems {
    let mut items = PrintItems::new();
    items.push_signal(Signal::StartForceNoNewLines);
    items.extend(item);
    items.push_signal(Signal::FinishForceNoNewLines);
    items
}

pub fn new_line_group(item: PrintItems) -> PrintItems {
    let mut items = PrintItems::new();
    items.push_signal(Signal::StartNewLineGroup);
    items.extend(item);
    items.push_signal(Signal::FinishNewLineGroup);
    items
}

// todo: move these conditions to the conditions module

pub fn if_true(
    name: &'static str,
    resolver: impl Fn(&mut ConditionResolverContext) -> Option<bool> + Clone + 'static,
    true_path: PrintItems
) -> Condition {
    Condition::new(name, ConditionProperties {
        true_path: Some(true_path),
        false_path: None,
        condition: Box::new(resolver.clone()),
    })
}

pub fn if_true_or(
    name: &'static str,
    resolver: impl Fn(&mut ConditionResolverContext) -> Option<bool> + Clone + 'static,
    true_path: PrintItems,
    false_path: PrintItems
) -> Condition {
    Condition::new(name, ConditionProperties {
        true_path: Some(true_path),
        false_path: Some(false_path),
        condition: Box::new(resolver.clone())
    })
}

pub fn if_false(
    name: &'static str,
    resolver: impl Fn(&mut ConditionResolverContext) -> Option<bool> + Clone + 'static,
    false_path: PrintItems
) -> Condition {
    Condition::new(name, ConditionProperties {
        true_path: None,
        false_path: Some(false_path),
        condition: Box::new(resolver.clone()),
    })
}

/// Parses a string as is and ignores its indent.
pub fn parse_raw_string(text: &str) -> PrintItems {
    let add_ignore_indent = text.find("\n").is_some();
    let mut items = PrintItems::new();
    if add_ignore_indent { items.push_signal(Signal::StartIgnoringIndent); }
    items.extend(parse_string(text));
    if add_ignore_indent { items.push_signal(Signal::FinishIgnoringIndent); }

    return items;
}

/// Parses a string to a series of PrintItems.
pub fn parse_string(text: &str) -> PrintItems {
    let mut items = PrintItems::new();
    let mut lines = text.lines().collect::<Vec<&str>>();

    // todo: this is kind of hacky...
    // using .lines() will remove the last line, so add it back if it exists
    if text.ends_with("\n") {
        lines.push("");
    }

    for i in 0..lines.len() {
        if i > 0 {
            items.push_signal(Signal::NewLine);
        }

        items.extend(parse_line(&lines[i]));
    }

    return items;

    fn parse_line(line: &str) -> PrintItems {
        let mut items = PrintItems::new();
        let parts = line.split("\t").collect::<Vec<&str>>();
        for i in 0..parts.len() {
            if i > 0 {
                items.push_signal(Signal::Tab);
            }
            if !parts[i].is_empty() {
                items.push_str(parts[i]);
            }
        }
        items.into()
    }
}

/// Surrounds the items with newlines and indentation if its on multiple lines.
/// Note: This currently inserts a possible newline at the start, but that might change or be made
/// conditional in the future.
pub fn surround_with_newlines_indented_if_multi_line(inner_items: PrintItems, indent_width: u8) -> PrintItems {
    let mut items = PrintItems::new();
    let start_info = Info::new("surroundWithNewLinesIndentedIfMultiLineStart");
    let end_info = Info::new("surroundWithNewLineIndentedsIfMultiLineEnd");
    let inner_items = inner_items.into_rc_path();

    items.push_info(start_info);
    items.push_condition(Condition::new_with_dependent_infos("newlineIfMultiLine", ConditionProperties {
        true_path: Some(surround_with_new_lines(with_indent(inner_items.clone().into()))),
        false_path: Some({
            let mut items = PrintItems::new();
            items.push_condition(super::conditions::if_above_width(
                indent_width,
                Signal::PossibleNewLine.into()
            ));
            items.extend(inner_items.into());
            items
        }),
        condition: Box::new(move |context| super::condition_resolvers::is_multiple_lines(context, &start_info, &end_info))
    }, vec![end_info]));
    items.push_info(end_info);

    items
}
